clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -527,4 +536,87 @@ public void onComplete(ClientResponse response) {
+            unsent = new HashMap<>();
+        }
+
+        public synchronized void put(Node node, List<ClientRequest> requests) {
+            List<ClientRequest> nodeUnsent = unsent.get(node);
+            if (nodeUnsent == null) {
+                nodeUnsent = new ArrayList<>();
+                unsent.put(node, nodeUnsent);
+            }
+            nodeUnsent.addAll(requests);
+        }
+
+        public synchronized void put(Node node, ClientRequest request) {
+            List<ClientRequest> nodeUnsent = unsent.get(node);
+            if (nodeUnsent == null) {
+                nodeUnsent = new ArrayList<>();
+                unsent.put(node, nodeUnsent);
+            }
+            nodeUnsent.add(request);
+        }
+
+        public synchronized int getRequestCount(Node node) {
+            List<ClientRequest> requests = unsent.get(node);
+            return requests == null ? 0 : requests.size();
+        }
+
+        public synchronized int getRequestCount() {
+            int total = 0;
+            for (List<ClientRequest> requests : unsent.values())
+                total += requests.size();
+            return total;
+        }
+
+        public synchronized boolean hasRequest(Node node) {
+            List<ClientRequest> requests = unsent.get(node);
+            return requests != null && !requests.isEmpty();
+        }
+
+        public synchronized boolean hasRequest() {
+            for (List<ClientRequest> requests : unsent.values())
+                if (!requests.isEmpty())
+                    return true;
+            return false;
+        }
+
+        public synchronized List<ClientRequest> removeExpiredRequests(long now, long unsentExpiryMs) {
+            List<ClientRequest> expiredRequests = new ArrayList<>();
+            Iterator<Map.Entry<Node, List<ClientRequest>>> iterator = unsent.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<Node, List<ClientRequest>> requestEntry = iterator.next();
+                Iterator<ClientRequest> requestIterator = requestEntry.getValue().iterator();
+                while (requestIterator.hasNext()) {
+                    ClientRequest request = requestIterator.next();
+                    if (request.createdTimeMs() < now - unsentExpiryMs) {
+                        expiredRequests.add(request);
+                        requestIterator.remove();
+                    } else
+                        break;
+                }
+                if (requestEntry.getValue().isEmpty())
+                    iterator.remove();
+            }
+            return expiredRequests;
+        }
+
+        public synchronized List<ClientRequest> remove(Node node) {
+            List<ClientRequest> requests = unsent.remove(node);
+            return requests == null ? Collections.<ClientRequest>emptyList() : requests;
+        }
+
+        public synchronized Set<Node> getNodes() {
+            return new HashSet<>(unsent.keySet());
```

Comment:
It's a bit annoying that we create a copy every time we call this method. It would be nicer if we only did that when it was required (e.g. connection failed or client is ready). If that's too complicated, we can at least return an `ArrayList` and save the cost of `HashSet` since all we do is iterate every time.